Document contributor process for adding witness hint templates (#933)
obi1kenobi/cargo-semver-checks: CONTRIBUTING.md

Congrats on the new lint!

#### Adding a witness

> **Witnesses** are a new, unstable feature of `cargo-semver-checks` that let us create witness
> code, a minimal compilable example of potential downstream breakage\. They are configured via the

Reviewer: Let's explain the new term directly instead of making a circular definition first.
```suggestion
**Witnesses** are a new, unstable feature of `cargo-semver-checks` that let us create
minimal compile-able examples of potential downstream breakage\. They are configured via the
```

> code, a minimal compilable example of potential downstream breakage\. They are configured via the

Author: Just a bad regex find-and-replace for replacing `.  ` with `. ` (extra spaces between periods). :sweat_smile:

> code, a minimal compilable example of potential downstream breakage\. They are configured via the

Reviewer: Ah okay, no worries. You might need to edit my suggestions post-merge then, since I mirrored the `\.` style.

`witness` field in the lint file `src/lints/<lint_name>.ron`:

If it is `None` (or the field is omitted entirely), `cargo-semver-checks` will not be able
to generate witness code for this lint\. This can be the right choice, sometimes:
for example, if this lint is a warn- or allow-by-default lint that hints at potential
breakage, but won't cause breaking changes directly\. Additionally, if it's not currently
possible to write a Trustfall query that gets the necessary information to generate
witnesses, leave this field as `None`, but leave a `// TODO` comment explaining
what would unblock this lint from being able to generate a witness.

##### Hint templates

> When the `witness` field is not `None`, it must have the `hint_template` field\. This is a
> `handlebars` template that generates a small (1-3 line) human-readable message that
> explains the idea of how downstream code would break\. For example, for the `function_missing` lint:

Reviewer: ```suggestion
When the `witness` field is not `None`, it must have the `hint_template` field\. This is a
`handlebars` template that generates a small (1-3 line) human-readable message that
explains the idea of how downstream code would break.

This example code is meant to be small and illustrative, and does not have to pass `cargo check`. It should give the reader a sense of the kind of breakage in one glance.

For example, for the `function_missing` lint, a witness template may look like this:
```

```rust

> use {{join "::" path};
> {{name}}(...);

Reviewer: This is actually broken in the same way the enum variants example was, it will fail under renaming =/

```suggestion
{{join "::" path}(...);
```

Author: Will fix in #934

```

> which could render to something like:

Reviewer: "Render" is a bit ambiguous here.
```suggestion
which in the RON format could be represented like:
```

```ron
witness: (
  hint_template: r#"use function_missing::will_be_removed_fn;\n\
  will_be_removed_fn(...);"#
)

> ```

Reviewer: It might be worth illustrating multi-line templates in a separate example, after going through the basic example end-to-end first.

> A witness hint like this may not be buildable (e.g., we elide the function arguments here,
> and the function call is not inside a block), but a witness hint should be a distilled
> example of breakage, and should not require extra information beyond the original
> lint's query.

Reviewer: ```suggestion
This hint will not pass `cargo check` (e.g. the function call arguments are elided), 
and that's okay. The hint is a distilled example of breakage, and shouldn't require
additional information beyond what the lint query retrieved.
```

##### Templating

We use the `handlebars` crate for writing these templates\. [More information

> more information about the syntax can be found here](https://handlebarsjs.com/guide/#simple-expressions),

Reviewer: duplicated words
```suggestion
about the syntax can be found here](https://handlebarsjs.com/guide/#simple-expressions),
```

and [here is where `cargo-semver-checks` defines custom helpers
](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/src/templating.rs).

All fields marked with `@output` in the `query` in `<lint_name>.ron` are available
to access with `{{name}}` in the `hint_template`, like in the example above.

> ##### Testing

Reviewer: ```suggestion
##### Testing witnesses
```

        ),
        (
            filename: "src/lib.rs",
            begin_line: 4,
            hint: "use function_missing::pub_use_removed_fn;\npub_use_removed_fn(...);",
        ),
    ],
}
thread 'query::tests_lints::function_missing' panicked at src/query.rs:751:17:
Witness output for function_missing did not match expected values:
Differences (-expected|+actual):
-{}
+{
+    // cut down for readibility
+    "./test_crates/function_missing/": [
+        (
+            filename: "src/lib.rs",
+            begin_line: 1,
+            hint: "use function_missing::will_be_removed_fn;\nwill_be_removed_fn(...);",
+        ),
+        (
+            filename: "src/lib.rs",
+            begin_line: 4,
+            hint: "use function_missing::pub_use_removed_fn;\npub_use_removed_fn(...);",
+        ),
+    ],
+}

Update the `test_outputs/witnesses/function_missing.output.ron` file if
                    the new test results are correct.
```

Check the actual result under `--- actual ---` and make sure that it outputted
the correct witness hints\. Note that it may contain output for other test crates - this

> not necessarily an error: see the troubleshooting section for more info\. Once you've

Reviewer: ```suggestion
is not necessarily an error: see the troubleshooting section for more info\. Once you've
```

verified that the `--- actual ---` results are as expected, copy the actual results

> into the `test_outputs/witnesses/<lint_name>.output.ron` file and save\. Running

Reviewer: Is there a viable way to `insta` this? 🤔

Author: Totally! I was matching the behavior of the basic lint tests, but we could definitely use `insta` here, and it would probably be better for contributor experience (but less consistent)

Reviewer: I imagine we'll eventually move the existing non-insta steps to insta, so probably best to start standardizing toward the new behavior we'll want eventually anyway. If someone is annoyed at the inconsistency, we can tell them we'd love a PR to move everything to insta 😅

Author: OK, I'll open a PR to use insta for the tests instead, then update the docs here.

Author: #935

Congrats on the new lint!

#### Adding a witness

**Witnesses** are a new, unstable feature of `cargo-semver-checks` that let us create
minimal compile-able examples of potential downstream breakage. They are configured via the
`witness` field in the lint file `src/lints/<lint_name>.ron`:

If it is `None` (or the field is omitted entirely), `cargo-semver-checks` will not be able
to generate witness code for this lint. This can be the right choice, sometimes:
for example, if this lint is a warn- or allow-by-default lint that hints at potential
breakage, but won't cause breaking changes directly. Additionally, if it's not currently
possible to write a Trustfall query that gets the necessary information to generate
witnesses, leave this field as `None`, but leave a `// TODO` comment explaining
what would unblock this lint from being able to generate a witness.

##### Hint templates

When the `witness` field is not `None`, it must have the `hint_template` field. This is a
`handlebars` template that generates a small (1-3 line) human-readable message that
explains the idea of how downstream code would break.

> This example code is meant to be small and illustrative, and does not have to pass `cargo check`. It should give the reader a sense of the kind of breakage in one glance.

Reviewer: ```suggestion
This example code is meant to be small and illustrative, and does not have to pass `cargo check`.
It should give the reader a sense of the kind of breakage in one glance.
```
